File: 3.longest-substring-without-repeating-characters.rb
#
# @lc app=leetcode id=3 lang=ruby
#
# [3] Longest Substring Without Repeating Characters
#

# @lc code=start
# @param {String} s
# @return {Integer}
def length_of_longest_substring(s)
  hash = {}
  max = 0
  start = 0
  s.each_char.with_index do |c, i|
    start = hash[c] + 1 if hash[c] && hash[c] >= start
    max = [max, i - start + 1].max
    hash[c] = i
  end
  max
end
# @lc code=end

File: 3.longest-substring-without-repeating-characters.js
/*
 * @lc app=leetcode id=3 lang=javascript
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
  const map = new Map();
  let max = 0;
  for (let i = 0, j = 0; j < s.length; ++j) {
    if (map.has(s[j])) {
      i = Math.max(i, map.get(s[j]) + 1); // 这里易错 i=map.get(s[j])+1
    }
    map.set(s[j], j);
    max = Math.max(max, j - i + 1);
  }
  return max;
};
// @lc code=end

lengthOfLongestSubstring("abba");
